common/task/weighted_channel_pool.go

```diff
@@ -0,0 +1,225 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package task
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/clock"
+)
+
+const defaultIdleChannelTTLInSeconds = 3600
+
+type (
+	weightedChannels[V any] []*weightedChannel[V]
+
+	weightedChannel[V any] struct {
+		weight         int
+		c              chan V
+		refCount       atomic.Int32
+		lastAccessTime atomic.Int64
+	}
+
+	WeightedRoundRobinChannelPoolOptions struct {
+		BufferSize              int
+		IdleChannelTTLInSeconds int64
+	}
+
+	WeightedRoundRobinChannelPool[K comparable, V any] struct {
```

Comment:
Thanks for refactoring weighted-round-robin schduler by extracting the channel pool to a separate component and simplifying the implementation. This abstraction makes a lot of sense